core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *  
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *  
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *  
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.batch.file;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract class that contains file metadata fields.
+ * Extend from this class to add credentials specific to different filesystems.
+ */
+public abstract class AbstractFileMetadata {
+  public static final String FILE_NAME = "fileName";
+  public static final String FILE_SIZE = "fileSize";
+  public static final String TIMESTAMP = "timeStamp";
+  public static final String OWNER = "owner";
+  public static final String FULL_PATH = "fullPath";
+  public static final String IS_FOLDER = "isFolder";
+  public static final String BASE_PATH = "basePath";
+  public static final String PERMISSION = "permission";
+
+  // contains only the name of the file
+  protected String fileName;
+
+  // full path of the file in the source filesystem
+  protected String fullPath;
+
+  // file size
+  protected long fileSize;
+
+  // modification time of file
+  protected long timeStamp;
+
+  // owner of file
+  protected String owner;
+
+  // whether or not the file is a folder
+  protected Boolean isFolder;
+
+  // the base path that will be appended to the path the sink is writing to
+  protected String basePath;
+
+  // perission of file, encoded in short
+  protected short permission;
+
+  // Credentials needed to connect to the filesystem that contains this file
+  protected final Credentials credentials;
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractFileMetadata.class);
+
+  public AbstractFileMetadata(FileStatus fileStatus, String sourcePath, Credentials credentials) {
+    String fullPath = fileStatus.getPath().toString();
+    String[] paths = fullPath.split("/");
+    this.fileName = paths[paths.length - 1];
+    this.timeStamp = fileStatus.getModificationTime();
+    this.owner = fileStatus.getOwner();
+    this.fileSize = fileStatus.getLen();
+    this.credentials = credentials;
+    this.permission = fileStatus.getPermission().toShort();
+
+    // generate file folder
+    this.fullPath = fullPath;
+
+    // check if file is a folder
+    this.isFolder = fileStatus.isDirectory();
+
+    // TODO: investigate how to cleanly set basePath
+    if (sourcePath.equals("")) {
+      basePath = this.fileName;
```

Comment:
Example: suppose there is a file in /a/b/c/d/haha.txt
If you do `cp  -r   /a/b/c   /dest/`
Base path in this case would be 'c/d/haha.txt'. Therefore the resulting file path would be `/dest/c/d/haha.txt`

I'll add a comment in the code to explain it more clearly.